src/Func.h

```diff
@@ -714,18 +715,62 @@ class Func {
      Buffer<float> im1 = r[1];
      \endcode
      *
+     * In Halide formal arguments of a computation are specified using
+     * Param<T> and ImageParam objects in the expressions defining the
+     * computation. The param_map argument to realize allows
+     * specifying a set of per-call parameters to be used for a
+     * specific computation. This method is thread-safe where the
+     * globals used by Param<T> and ImageParam are not. Any parameters
+     * that are not n the param_map are taken from the global values,
+     * so those can continue to be used if they are not changing
+     * per-thread.
+     *
+     * One can explicitly construct a ParamMap and
+     * use its set method to insert Parameter to scalar or Buffer
+     * value mappings:
+     *
+     \code
+     Param<int32> p(42);
+     ImageParam img(Int(32), 1);
+     f(x) = img(x) + p;
+
+     Buffer<int32_t) arg_img(10, 10);
+     <fill in arg_img...>
+     ParamMap params;
+     params.set(p, 17);
+     params.set(img, arg_img);
+
+     Target t = get_jit_target_from_environment();
+     Buffer<int32_t> result = f.realize(10, 10, t, params);
+     \endcode
+     *
+     * Alternatively, an initializer list can be used
+     * directly in the realize call to pass this information:
+     *
+     \code
+     Param<int32> p(42);
+     ImageParam img(Int(32), 1);
+     f(x) = img(x) + p;
+
+     Buffer<int32_t) arg_img(10, 10);
+     <fill in arg_img...>
+
+     Target t = get_jit_target_from_environment();
+     Buffer<int32_t> result = f.realize(10, 10, t, { { p, 17 }, { img, arg_img } });
+     \endcode
+     *
      */
     // @{
-    EXPORT Realization realize(std::vector<int32_t> sizes, const Target &target = Target());
+    EXPORT Realization realize(std::vector<int32_t> sizes, const Target &target = Target(), const ParamMap &param_map = ParamMap());
```

Comment:
I realize there is no other good way to do this without breaking the API, but methods with multiple default value arguments make my spidey-sense tingle. 

Since this is an unusual use case, and since all but one of the realize() methods are just convenience wrappers, would it make sense to offer ParamsMap on just that underlying one, to avoid clogging the API?

Reply:
As much as I don't like the multiple default param thing, I also don't like variants of a method having different capabilities as I find it leads to a fair bit of confusion in actual use. One has three calls to the same conceptual method and then has to convert the convenience versions to the primitive version to make a code change. At that point the convenience versions are no longer actually delivering convenience.

There are a few ways we could use C++ fu to make these sorts of methods work a bit better. Things that either use a cascade of method invocations to build up the call or things that use an initializer list of typed things to fill in a struct of all the optional/varying params to a complicated call. Not sure if we want to go down that path, but it is a possibility. It would likely want to be part of a incompatible API revamp.

Reply:
Decided to just go with multiple default args. It is not ideal, but other approaches somewhat fall down as the target and param map arguments are different in that target is used during compilation and param map is only used for the call. One approach would be to force separate compilation for folks using param maps, but that ends up replicating a bunch of calls and doesn't really make anything simpler.